Add depends 32-bit arm support for bitcoin-qt

sedited opened the pull request:
Some hobbyists are used to using the desktop for interfacing with their raspberry pi. This commits adds qt to the arm-linux-gnueabihf target.

Sjors commented on depends/packages/packages.mk:
```diff
@@ -5,6 +5,7 @@ qt_packages = qrencode protobuf zlib
 
 qt_x86_64_linux_packages:=qt expat dbus libxcb xcb_proto libXau xproto freetype fontconfig libX11 xextproto libXext xtrans
 qt_i686_linux_packages:=$(qt_x86_64_linux_packages)
+qt_arm_linux_packages:=$(qt_x86_64_linux_packages) $(qt_packages)
```
What's the role of `$(qt_packages)`, especially given that `qt_i686_linux_packages` doesn't have it?

TheCharlatan replied:
Good findings. I could remove all these changes (I had them in while I was debugging things) and it still seems to run fine. Can you re-run a gitian with b86d49f  ?

Sjors commented on depends/packages/qt.mk:
```diff
@@ -64,7 +64,7 @@ $(package)_config_opts += -prefix $(host_prefix)
 $(package)_config_opts += -qt-libjpeg
 $(package)_config_opts += -qt-pcre
-$(package)_config_opts += -system-zlib
+$(package)_config_opts += -qt-zlib
```
For future reference: what's wrong with `system-zlib`?

Sjors commented on depends/packages/qt.mk:
```diff
@@ -155,6 +157,9 @@ define $(package)_config_cmds
   cd ../lupdate/ && ../../../../qtbase/bin/qmake lupdate.pro -o Makefile && cd ../../../..
 endef
 
+#echo "host_build: QT_CONFIG ~= s/system-zlib/zlib" >> mkspecs/qconfig.pri && \
```
Should this go away?

Sjors reviewed (commented):
Concept ACK. Not just Raspberry Pi, also various clones via Armbian.

Fixes the 32-bit part of #13495.

TheCharlatan commented:
The Travis job times out with this change. Not sure how to get it to run through.

Sjors commented:
I restarted the Travis job, but that doesn't seem to help much.

Sjors commented:
I'm able to build a binary using Armbian: https://github.com/Sjors/armbian-bitcoin-core/pull/1

However, when I launch QT on an Orange Pi I get:

```
bitcoin-qt: /lib/arm-linux-gnueabihf/libm.so.6: version `GLIB_2.27` not found (required by bitcoin-qt)
```

This might be because I compiled on Bionic and the target machine uses Xenial, with libc6 version 2.23

I'll try again with an Armbian on a Ubuntu 16.04 VM.

TheCharlatan commented:
Alternatively you could also apply #13177 and test if it also works for arm.

Sjors commented:
Cherry-picking that branch did the trick, nice!

Sjors commented:
I'm also trying a gitian build, but that gets stuck at something that seems pretty unrelated (not even the same architecture):
```
  CXX      crypto/crypto_libbitcoin_crypto_sse41_a-sha256_sse41.o
  CXX      crypto/crypto_libbitcoin_crypto_avx2_a-sha256_avx2.o
crypto/sha256.cpp: In function ‘std::string SHA256AutoDetect()’:
crypto/sha256.cpp:537:83: error: inconsistent operand constraints in an ‘asm’
   __asm__ ("cpuid" : "=a"(a), "=b"(b), "=c"(c), "=d"(d) : "0"(leaf), "2"(subleaf));
                                                                                   ^
crypto/sha256.cpp:537:83: error: inconsistent operand constraints in an ‘asm’
   __asm__ ("cpuid" : "=a"(a), "=b"(b), "=c"(c), "=d"(d) : "0"(leaf), "2"(subleaf));
                                                                                   ^
make[2]: *** [crypto/crypto_libbitcoin_crypto_base_a-sha256.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/home/ubuntu/build/bitcoin/distsrc-i686-pc-linux-gnu/src'
```

ken2812221 commented:
@Sjors See #13538

Sjors commented:
@ken2812221 thanks. I just deleted those architectures from the gitian descriptor and tried again.

```
9c0a172871a27bb727ed27fcdf2f5fd3c6e6b5bc82ea6d7b5745916887a450cb  bitcoin-0.16.99-arm-linux-gnueabihf.tar.gz
```

Binary works on device as well.

tACK 8ef00e8, assuming Travis gives its blessing

laanwj commented:
my first question when seeing this is: why only 32-bit?—is this already supported for 64-bit, or does 64-bit ARM provide additional challenges?

TheCharlatan commented:
@laanwj see my comment in #13495 on the difficulties I am having with 64bit arm.

laanwj commented:
> @laanwj see my comment in #13495 on the difficulties I am having with 64bit arm.

Fair enough. It's fine, we can do that in a later PR.

Looks like the travis issue is simply a timeout running into the maximum time limit for jobs.

Sjors commented:
A particularly stubborn Travis timeout. Maybe the ccache needs to be bigger?

TheCharlatan commented:
I'm not sure if your suggestion @Sjors may have unwanted side effects to the Travis jobs. It seems like the linux qt builds are disabled for this reason in the Travis build matrix. Otherwise I'll remove the test from the pr and hope that enough people run it through gitian to verify?

TheCharlatan commented:
Removed the Travis patch. To test this pull request, build it locally.

Sjors commented:
I restarted the Apple Darwin instance

theuni commented on depends/packages/packages.mk:
```diff
@@ -5,6 +5,7 @@ qt_packages = qrencode protobuf zlib
 
 qt_x86_64_linux_packages:=qt expat dbus libxcb xcb_proto libXau xproto freetype fontconfig libX11 xextproto libXext xtrans
 qt_i686_linux_packages:=$(qt_x86_64_linux_packages)
+qt_arm_linux_packages:=$(qt_x86_64_linux_packages)
```
Since there's no point in making the distinction here, let's just get rid of it.

```qt_x86_64_linux_packages``` can just become ```qt_linux_packages```, then ``` qt_i686_linux_packages``` and ```qt_arm_linux_package``` are no longer needed.

TheCharlatan replied:
The patch does not work yet for aarch64, so I think we should keep it like this until that it is fixed.

theuni replied:
Making the suggested change above will make it so that no change is needed here for aarch64.

TheCharlatan replied:
Substituting `qt_x86_64_linux_packages` with `qt_linux_packages` means that qt will be built for all Linux architectures. When I do that and remove the two lines below, a `aarch64-linux-gnu` depends build will also try to build qt, but will fail as I mentioned in #13495 .

theuni replied:
Yes, ok, I forgot we purposefully limited that way.

theuni reviewed (changes requested):
Concept ACK.

TheCharlatan commented on depends/packages/qt.mk:
```diff
@@ -89,7 +89,7 @@ $(package)_config_opts_linux += -system-freetype
 $(package)_config_opts_linux += -fontconfig
 $(package)_config_opts_linux += -no-opengl
-$(package)_config_opts_arm_linux  = -platform linux-g++ -xplatform $(host)
+$(package)_config_opts_arm_linux  = -platform linux-g++ -xplatform linux-arm-gnueabihf-g++
```
The sed patch is only intended for adding hard float support. I don't think this should be done for aarch64 as well (which seems to be the only other mkspec that matches the $(host) triplet).

theuni replied:
Yes, but then we can reuse it for aarch64. All that would be necessary to add it (assuming everything else is working) would be:
```
$(package)_config_opts_aarch64_linux = -platform linux-g++ -xplatform $(host)-g++
```

TheCharlatan replied:
I get `Invalid target platform 'arm-linux-gnueabihf-g++'` when I replace it with `$(host)-g++`.

theuni replied:
This works for me:

```patch
diff --git a/depends/packages/qt.mk b/depends/packages/qt.mk
index b4d7756..c09c02d 100644
--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -89,7 +89,7 @@ $(package)_config_opts_linux += -system-freetype
 $(package)_config_opts_linux += -no-feature-sessionmanager
 $(package)_config_opts_linux += -fontconfig
 $(package)_config_opts_linux += -no-opengl
-$(package)_config_opts_arm_linux  = -platform linux-g++ -xplatform $(host)
+$(package)_config_opts_arm_linux += -platform linux-g++ -xplatform $(host)-g++
 $(package)_config_opts_i686_linux  = -xplatform linux-g++-32
 $(package)_config_opts_x86_64_linux = -xplatform linux-g++-64
 $(package)_config_opts_mingw32  = -no-opengl -xplatform win32-g++ -device-option CROSS_COMPILE="$(host)-"
@@ -128,6 +128,8 @@ define $(package)_preprocess_cmds
   cp -f qtbase/mkspecs/macx-clang/Info.plist.app qtbase/mkspecs/macx-clang-linux/ &&\
   cp -f qtbase/mkspecs/macx-clang/qplatformdefs.h qtbase/mkspecs/macx-clang-linux/ &&\
   cp -f $($(package)_patch_dir)/mac-qmake.conf qtbase/mkspecs/macx-clang-linux/qmake.conf && \
+  cp -r qtbase/mkspecs/linux-arm-gnueabi-g++ qtbase/mkspecs/$(host)-g++ && \
+  sed -i s/arm-linux-gnueabi-/$(host)-/g qtbase/mkspecs/$(host)-g++/qmake.conf && \
   patch -p1 -i $($(package)_patch_dir)/fix_qt_pkgconfig.patch &&\
   patch -p1 -i $($(package)_patch_dir)/fix_configure_mac.patch &&\
   patch -p1 -i $($(package)_patch_dir)/fix_no_printer.patch &&\
```

TheCharlatan replied:
Just noticed that the mkspecs is not `arm-linux-gnu` (like $(host)), but rather linux-arm-gnu. :p

TheCharlatan replied:
The second chunk could result in undesired behaviour, because you end up copying the arm mkspecs over whatever the $(host)'s mkspecs are supposed to be.

theuni replied:
Good point. Using ```bitcoin-$(host)-g++``` would guarantee that can't happen.

TheCharlatan replied:
I do not understand how that would work. Can you post another patch for it?

theuni replied:
I guess the spec name doesn't need to change, all that matters is the sed. Simplified:
```patch
diff --git a/depends/packages/qt.mk b/depends/packages/qt.mk
index b4d7756..22cf4f5 100644
--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -89,7 +89,7 @@ $(package)_config_opts_linux += -system-freetype
 $(package)_config_opts_linux += -no-feature-sessionmanager
 $(package)_config_opts_linux += -fontconfig
 $(package)_config_opts_linux += -no-opengl
-$(package)_config_opts_arm_linux  = -platform linux-g++ -xplatform $(host)
+$(package)_config_opts_arm_linux += -platform linux-g++ -xplatform bitcoin-linux-g++
 $(package)_config_opts_i686_linux  = -xplatform linux-g++-32
 $(package)_config_opts_x86_64_linux = -xplatform linux-g++-64
 $(package)_config_opts_mingw32  = -no-opengl -xplatform win32-g++ -device-option CROSS_COMPILE="$(host)-"
@@ -128,6 +128,8 @@ define $(package)_preprocess_cmds
   cp -f qtbase/mkspecs/macx-clang/Info.plist.app qtbase/mkspecs/macx-clang-linux/ &&\
   cp -f qtbase/mkspecs/macx-clang/qplatformdefs.h qtbase/mkspecs/macx-clang-linux/ &&\
   cp -f $($(package)_patch_dir)/mac-qmake.conf qtbase/mkspecs/macx-clang-linux/qmake.conf && \
+  cp -r qtbase/mkspecs/linux-arm-gnueabi-g++ qtbase/mkspecs/bitcoin-linux-g++ && \
+  sed -i s/arm-linux-gnueabi-/$(host)-/g qtbase/mkspecs/bitcoin-linux-g++/qmake.conf && \
   patch -p1 -i $($(package)_patch_dir)/fix_qt_pkgconfig.patch &&\
   patch -p1 -i $($(package)_patch_dir)/fix_configure_mac.patch &&\
   patch -p1 -i $($(package)_patch_dir)/fix_no_printer.patch &&\
```
The effect is that we essentially have a universal linux mkspec now, it just changes the compiler names to match the HOST passed in.

TheCharlatan replied:
ACK.

sedited committed:
Add depends 32-bit arm support for bitcoin-qt

Some hobbyists are used to using the desktop for interfacing with their
raspberry pi. This commits adds qt to the arm-linux-gnueabihf target.

TheCharlatan commented:
Force-pushed to respond to @theuni's review.

Fuzzbawls commented:
Concept ACK

aarch64 support can be added later, after the issue with xextproto are dealt with and migration to Bionic for gitian builds is complete.

Sjors commented:
re-tACK 4b69984 (I retested cross-compile and running on a device, not Gitian)

TheCharlatan commented:
I get the same xml error when I try to apply @Fuzzbawls patch for aarch64 support.

sipa commented:
@theuni Does your "changes requested" still apply?

theuni reviewed (approved):
utACK 4b6998455776ed3e1b7c996c14f15f0c304c9269

Sjors commented:
Maybe add "for cross-compilation" to the commit message?

ken2812221 commented:
utACK 4b69984

Sjors commented:
Gitian builds for master are now broken: #13700

Sjors commented on depends/packages/qt.mk:
```diff
@@ -128,6 +128,8 @@ define $(package)_preprocess_cmds
   cp -f qtbase/mkspecs/macx-clang/qplatformdefs.h qtbase/mkspecs/macx-clang-linux/ &&\
   cp -f $($(package)_patch_dir)/mac-qmake.conf qtbase/mkspecs/macx-clang-linux/qmake.conf && \
+  cp -r qtbase/mkspecs/linux-arm-gnueabi-g++ qtbase/mkspecs/bitcoin-linux-g++ && \
+  sed -i s/arm-linux-gnueabi-/$(host)-/g qtbase/mkspecs/bitcoin-linux-g++/qmake.conf && \
```
macOS doesn't like this, see #13750

hebasto commented:
Please see #18536.
